Modules/Quest/QuestEventHandler.lua

```diff
@@ -0,0 +1,352 @@
+    local questObjectives = C_QuestLog.GetQuestObjectives(questId)
+    for _, objective in pairs(questObjectives) do
+        -- When the objective text is not cached yet it looks similar to " slain 0/1"
+        if (not objective.text) or stringSub(objective.text, 1, 1) == " " then
+            print("--> Objective texts are not correct yet")
+            _QuestLogUpdateQueue:Insert(function()
+                return _QuestEventHandler:HandleQuestAccepted(questId)
+            end)
+
+            objectiveTextTries = objectiveTextTries + 1
+            -- No need to check other objectives since we have to check them all again already
+            return false
+        end
+    end
+
+    print("--> Objectives are correct. Calling accept logic. quest:", questId)
+    questLog[questId].state = QUEST_LOG_STATES.QUEST_ACCEPTED
+    QuestieHash:AddNewQuestHash(questId)
+    QuestieQuest:AcceptQuest(questId)
+    QuestieJourney:AcceptQuest(questId)
+
+    objectiveTextTries = 0
+    return true
+end
+
+--- Fires when a quest is turned in
+---@param questId number
+---@param xpReward number
+---@param moneyReward number
+function _QuestEventHandler:QuestTurnedIn(questId, xpReward, moneyReward)
+    print("[Quest Event] QUEST_TURNED_IN", xpReward, moneyReward, questId)
+
+    if questLog[questId] and questLog[questId].timer then
+        -- Cancel the timer so the quest is not marked as abandoned
+        questLog[questId].timer:Cancel()
+        questLog[questId].timer = nil
+    end
+
+    print("--> Quest", questId, "was turned in and is completed")
+    if questLog[questId] then
+        -- There are quests which you just turn in so there is no preceding QUEST_ACCEPTED event and questLog[questId]
+        -- is empty
+        questLog[questId].state = QUEST_LOG_STATES.QUEST_TURNED_IN
+    end
+
+    local parentQuest = QuestieDB.QueryQuestSingle(questId, "parentQuest")
+    if parentQuest then
+        -- Quests like "The Warsong Reports" have child quests which are just turned in. These child quests only
+        -- fire QUEST_TURNED_IN + QUEST_LOG_UPDATE
+        doFullQuestLogScan = true
+    end
+
+    skipNextUQLCEvent = true
+    QuestieHash:RemoveQuestHash(questId)
+    QuestieQuest:CompleteQuest(questId)
+    QuestieJourney:CompleteQuest(questId)
+end
+
+--- Fires when a quest is removed from the quest log. This includes turning it in and abandoning it.
+---@param questId number
+function _QuestEventHandler:QuestRemoved(questId)
+    print("[Quest Event] QUEST_REMOVED", questId)
+
+    if (not questLog[questId]) then
+        print("--> questLog[questId] was nil")
+        questLog[questId] = {}
+    end
+
+    -- QUEST_TURNED_IN was called before QUEST_REMOVED --> quest was turned in
+    if questLog[questId].state == QUEST_LOG_STATES.QUEST_TURNED_IN then
+        print("--> Quest", questId, "was turned in before. Nothing do to.")
+        return
```

Comment:
Should also here be `skipNextUQLCEvent = true` as in other return branch? I don't know.